StreamProcessorReprocessingTest.shouldStartFromLastProcessedEvent flaky

**Summary**

- How often does the test fail? at least once
- Does it block your work? no
- Do we suspect that it is a real failure? unknown

**Failures**

<details><summary>Example assertion failure</summary>
<pre>
java.lang.AssertionError: 

Expecting:
  <[2L, 3L]>
to contain exactly (and in same order):
  <[3L]>
but some elements were not found:
  <[3L]>
and others were not expected:
  <[2L]>

	at io.zeebe.engine.processor.StreamProcessorReprocessingTest.shouldStartFromLastProcessedEvent(StreamProcessorReprocessingTest.java:366)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:27)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.apache.maven.surefire.junitcore.JUnitCore.run(JUnitCore.java:55)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.createRequestAndRun(JUnitCoreWrapper.java:137)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.executeLazy(JUnitCoreWrapper.java:119)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:87)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:75)
	at org.apache.maven.surefire.junitcore.JUnitCoreProvider.invoke(JUnitCoreProvider.java:157)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:428)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:162)
	at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:562)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:548)
</pre>
</details>

**Hypotheses**

**Logs**

<details><summary>Logs</summary>
<pre>
04:45:53.562 [] INFO  io.zeebe.test - Test finished: shouldIgnoreRecordWhenNoProcessorExistForThisType(io.zeebe.engine.processor.StreamProcessorReprocessingTest)
04:45:53.562 [] INFO  io.zeebe.test - Test started: shouldStartFromLastProcessedEvent(io.zeebe.engine.processor.StreamProcessorReprocessingTest)
04:45:53.633 [Broker-0-LogStream-0] DEBUG io.zeebe.logstreams - Configured log appender back pressure at partition 0 as AppenderVegasCfg{initialLimit=1024, maxConcurrency=32768, alphaLimit=0.7, betaLimit=0.95}. Window limiting is disabled
04:45:54.299 [Broker-0-StreamProcessor-0] DEBUG io.zeebe.logstreams - Recovering state of partition 0 from snapshot
04:45:54.300 [Broker-0-StreamProcessor-0] INFO  org.camunda.feel.FeelEngine - Engine created. [value-mapper: CompositeValueMapper(List(io.zeebe.el.impl.feel.MessagePackValueMapper@7bab722c)), function-provider: io.zeebe.el.impl.feel.FeelFunctionProvider@756c9329, configuration: Configuration(false)]
04:45:54.300 [Broker-0-StreamProcessor-0] INFO  io.zeebe.logstreams - Recovered state of partition 0 from snapshot at position -1
04:45:54.302 [] DEBUG io.zeebe.util.buffer - Close stream processor
04:45:54.303 [Broker-0-StreamProcessor-0] DEBUG io.zeebe.logstreams - Closed stream processor controller Broker-0-StreamProcessor-0.
04:45:54.341 [] INFO  io.zeebe.logstreams - Closed stream stream-0
04:45:54.454 [Broker-0-StreamProcessor-0] DEBUG io.zeebe.logstreams - Recovering state of partition 0 from snapshot
04:45:54.455 [Broker-0-StreamProcessor-0] INFO  org.camunda.feel.FeelEngine - Engine created. [value-mapper: CompositeValueMapper(List(io.zeebe.el.impl.feel.MessagePackValueMapper@136a6843)), function-provider: io.zeebe.el.impl.feel.FeelFunctionProvider@1dad945e, configuration: Configuration(false)]
04:45:54.458 [Broker-0-StreamProcessor-0] INFO  io.zeebe.logstreams - Recovered state of partition 0 from snapshot at position 1
04:45:54.465 [] INFO  io.zeebe.broker.test - Test failed, following records where exported:
04:45:54.467 [] INFO  io.zeebe.broker.test - Records on partition 0:
			Position: 1 Key: -1
				RecordMetadata{recordType=EVENT, intentValue=255, intent=ELEMENT_ACTIVATING, requestStreamId=-2147483648, requestId=-1, protocolVersion=1, valueType=WORKFLOW_INSTANCE, rejectionType=NULL_VAL, rejectionReason=}
				Value:
{"bpmnProcessId":"","version":-1,"workflowKey":-1,"workflowInstanceKey":1,"elementId":"","flowScopeKey":-1,"bpmnElementType":"UNSPECIFIED","parentWorkflowInstanceKey":-1,"parentElementInstanceKey":-1}
			Position: 2 Key: -1
				RecordMetadata{recordType=EVENT, intentValue=255, intent=ELEMENT_ACTIVATING, requestStreamId=-2147483648, requestId=-1, protocolVersion=1, valueType=WORKFLOW_INSTANCE, rejectionType=NULL_VAL, rejectionReason=}
				Value:
{"bpmnProcessId":"","version":-1,"workflowKey":-1,"workflowInstanceKey":1,"elementId":"","flowScopeKey":-1,"bpmnElementType":"UNSPECIFIED","parentWorkflowInstanceKey":-1,"parentElementInstanceKey":-1}
			Position: 3 Key: -1
				RecordMetadata{recordType=EVENT, intentValue=255, intent=ELEMENT_ACTIVATING, requestStreamId=-2147483648, requestId=-1, protocolVersion=1, valueType=WORKFLOW_INSTANCE, rejectionType=NULL_VAL, rejectionReason=}
				Value:
{"bpmnProcessId":"","version":-1,"workflowKey":-1,"workflowInstanceKey":1,"elementId":"","flowScopeKey":-1,"bpmnElementType":"UNSPECIFIED","parentWorkflowInstanceKey":-1,"parentElementInstanceKey":-1}

04:45:54.468 [] DEBUG io.zeebe.util.buffer - Close stream processor
04:45:54.468 [Broker-0-StreamProcessor-0] DEBUG io.zeebe.logstreams - Closed stream processor controller Broker-0-StreamProcessor-0.
04:45:54.504 [Broker-0-LogStream-0] INFO  io.zeebe.logstreams - Close appender for log stream stream-0
04:45:54.504 [stream-0-write-buffer] DEBUG io.zeebe.dispatcher - Dispatcher closed
04:45:54.504 [Broker-0-LogStream-0] INFO  io.zeebe.logstreams - On closing logstream stream-0 close 4 readers
04:45:54.504 [Broker-0-LogStream-0] INFO  io.zeebe.logstreams - Close log storage with name stream-0
04:45:54.504 [] DEBUG io.zeebe.broker.test - Clean up test files on path /tmp/junit10824872323323479201
</pre>
</details>

--- a/engine/src/test/java/io/zeebe/engine/processing/streamprocessor/StreamProcessorReprocessingTest.java
+++ b/engine/src/test/java/io/zeebe/engine/processing/streamprocessor/StreamProcessorReprocessingTest.java
@@ -316,9 +316,9 @@ public final class StreamProcessorReprocessingTest {
   }
 
   @Test
-  public void shouldStartFromLastProcessedEvent() throws Exception {
+  public void shouldStartAfterLastProcessedEventInSnapshot() throws Exception {
     // given
-    final CountDownLatch processingLatch = new CountDownLatch(2);
+    final CountDownLatch onProcessedListenerLatch = new CountDownLatch(2);
     streamProcessorRule.startTypedStreamProcessor(
         (processors, context) ->
             processors.onEvent(
@@ -331,17 +331,19 @@ public final class StreamProcessorReprocessingTest {
                       final TypedRecord<UnifiedRecordValue> record,
                       final TypedResponseWriter responseWriter,
                       final TypedStreamWriter streamWriter,
-                      final Consumer<SideEffectProducer> sideEffect) {
-                    processingLatch.countDown();
-                  }
-                }));
+                      final Consumer<SideEffectProducer> sideEffect) {}
+                }),
+        (t) -> onProcessedListenerLatch.countDown());
 
     streamProcessorRule.writeWorkflowInstanceEvent(ELEMENT_ACTIVATING);
-    streamProcessorRule.writeWorkflowInstanceEvent(ELEMENT_ACTIVATING);
-    processingLatch.await();
+    streamProcessorRule.writeWorkflowInstanceEvent(
+        ELEMENT_ACTIVATING); // should be processed and included in the snapshot
+    onProcessedListenerLatch.await();
     streamProcessorRule.closeStreamProcessor();
 
     // when
+    // The processor restarts with a snapshot that was the state of the processor before it
+    // was closed.
     final List<Long> processedPositions = new ArrayList<>();
     final CountDownLatch newProcessLatch = new CountDownLatch(1);
     streamProcessorRule.startTypedStreamProcessor(
@@ -372,7 +374,7 @@ public final class StreamProcessorReprocessingTest {
   @Test
   public void shouldNotReprocessEventAtLastProcessedEvent() throws Exception {
     // given
-    final CountDownLatch processingLatch = new CountDownLatch(3);
+    final CountDownLatch onProcessedListenerLatch = new CountDownLatch(2);
     streamProcessorRule.startTypedStreamProcessor(
         (processors, context) ->
             processors.onEvent(
@@ -385,18 +387,14 @@ public final class StreamProcessorReprocessingTest {
                       final TypedRecord<UnifiedRecordValue> record,
                       final TypedResponseWriter responseWriter,
                       final TypedStreamWriter streamWriter,
-                      final Consumer<SideEffectProducer> sideEffect) {
-                    processingLatch.countDown();
-                  }
-                }));
+                      final Consumer<SideEffectProducer> sideEffect) {}
+                }),
+        t -> onProcessedListenerLatch.countDown());
 
     streamProcessorRule.writeWorkflowInstanceEvent(ELEMENT_ACTIVATING, 1);
     final long snapshotPosition =
         streamProcessorRule.writeWorkflowInstanceEvent(ELEMENT_ACTIVATING, 1);
-    streamProcessorRule.writeWorkflowInstanceEvent(ELEMENT_ACTIVATING, 1);
-    // we write three event to make sure that the second event was committed and is part of the
-    // state (lastProcessedEvent)
-    processingLatch.await();
+    onProcessedListenerLatch.await();
     streamProcessorRule.closeStreamProcessor();
     final long lastSourceEvent =
         streamProcessorRule.writeWorkflowInstanceEvent(ELEMENT_ACTIVATING, 1);

